NFT/src/main.sw

```diff
@@ -0,0 +1,359 @@
+    /// Map((owner, operator) => approved)
+    operator_approval: StorageMap<(Identity,
+    Identity), bool>, /// The total number of tokens that have been minted.
+    /// This should only be incremented.
+    token_count: u64,
+
+    /// The total supply tokens that can be minted. Can only be set on the initalization of the
+    /// contract.
+    token_supply: u64,
+}
+
+impl NFT for Contract {
+    /// Constructor for the NFT. Calling this function will instantiate the `total_supply`, the `admin`
+    /// `Identity`, and the `access_control` boolean. These values can only be set once.
+    /// Before this function is called, the contract is unable to perform any minting or transfering of tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `admin` - The `Identity` which has the ability to mint if `access_control` is set to true and change the contract's admin.
+    /// * `access_control` - The `bool` which will determine whether identities will need to approval to mint.
+    /// * `token_supply` - The `u64` number representing the total supply of tokens which will be allowed to mint.
+    ///
+    /// # Reverts
+    ///
+    /// * When the constructor function has already been called.
+    /// * When the `token_supply` is set to 0.
+    /// * When `access_control` is set to true and no admin was given.
+    #[storage(read, write)]fn constructor(admin: Option<Identity>, access_control: bool, token_supply: u64) {
+        // This function can only be called once so if the token supply is already set it has
+        // already been called
+        require(storage.token_supply == 0, InitError::CannotReinitialize);
+        // The number of tokens that can be minted cannot be 0
+        require(token_supply != 0, InputError::TokenSupplyCannotBeZero);
+        // Access control is set to true but there was no admin given
+        require((access_control && admin.is_some()) || admin.is_none(), InitError::AccessControlSetAndAdminIsNone);
+
+        // Store the information given
+        storage.access_control = access_control;
+        storage.admin = admin;
+        storage.token_supply = token_supply;
+    }
+
+    /// Mints a specified amount of tokens to the given `to` `Identity`. Once a token has been minted,
+    /// it can be transfered and burned. Calling this mint function will increment the `total_count`.
+    /// If the NFT contract has not yet been initalized, any attempts to mint will fail as the
+    /// `total_supply` has not yet been set.
+    ///
+    /// # Arguments
+    ///
+    /// * `to` - The `Identity` which will own the minted tokens.
+    /// * `amount` - The `u64` number of tokens to be minted in this transaction.
+    ///
+    /// # Reverts
+    ///
+    /// * When the sender attempts to mint more tokens than total supply.
+    /// * When the sender is not the admin and `access_control` is set.
+    #[storage(read, write)]fn mint(to: Identity, amount: u64) {
```

Comment:
Maybe not 100% necessary in the initial version here, but an improvement would be to have both a `mint` and a `mint_batch` function.
This would help minimize the minting cost for those who only want to mint a single token.
Eliminating the loop for single mints may be a small gas saving, but these things have a way of adding up.

Reply:
I think this would be good as a new issue.